store/tikv/txn_scheduler.go

```diff
@@ -0,0 +1,189 @@
+	"github.com/pingcap/tidb/store/tikv/latch"
+	binlog "github.com/pingcap/tipb/go-binlog"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+type txnCommitter struct {
+	twoPC *twoPhaseCommitter
+	ch    chan error
+	ctx   context.Context
+	lock  *latch.Lock
+}
+
+func newTxnCommiter(ctx context.Context, committer *twoPhaseCommitter) *txnCommitter {
+	return &txnCommitter{
+		twoPC: committer,
+		ctx:   ctx,
+	}
+}
+
+func (tc *txnCommitter) commit(stale bool) uint64 {
+	if stale {
+		err := errors.Errorf("startTS %d is stale", tc.twoPC.startTS)
+		log.Debug(tc.twoPC.connID, err)
+		tc.ch <- errors.Annotate(err, txnRetryableMark)
+		return uint64(0)
+	}
+
+	commitTS, err := tc.execute()
+	tc.ch <- errors.Trace(err)
+	log.Debug(tc.twoPC.connID, " finish txn with startTS:", tc.twoPC.startTS, " commitTS:", commitTS, " error:", err)
+	return commitTS
+}
+
+func (tc *txnCommitter) execute() (commitTS uint64, err error) {
+	err = tc.twoPC.execute(tc.ctx)
+	if err != nil {
+		tc.twoPC.writeFinishBinlog(binlog.BinlogType_Rollback, 0)
+	} else {
+		commitTS = tc.twoPC.commitTS
+		tc.twoPC.writeFinishBinlog(binlog.BinlogType_Commit, int64(commitTS))
+	}
+	return
+}
+
+type txnMap struct {
+	txns map[uint64]*txnCommitter
+	sync.RWMutex
+}
+
+func (t *txnMap) put(startTS uint64, txn *txnCommitter) {
+	t.Lock()
+	defer t.Unlock()
+	if t.txns == nil {
+		t.txns = make(map[uint64]*txnCommitter)
+	}
+	if _, ok := t.txns[startTS]; ok {
+		panic(fmt.Sprintf("The startTS %d shouldn't be used in two transaction", txn.twoPC.startTS))
+	}
+	t.txns[startTS] = txn
+}
+
+func (t *txnMap) get(startTS uint64) *txnCommitter {
+	t.RLock()
+	defer t.RUnlock()
+	return t.txns[startTS]
+}
+
+func (t *txnMap) delete(startTS uint64) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.txns, startTS)
+}
+
+// txnScheduler is the scheduler to commit transactions.
+type txnScheduler struct {
+	latches *latch.Latches
+	txnMaps []txnMap
+}
+
+const (
+	txnSchedulerLatchesSize = 102400
+	txnSchedulerMapsCount   = 256
+)
+
+func newTxnScheduler(enableLatches bool) *txnScheduler {
+	if !enableLatches {
+		return &txnScheduler{
+			latches: nil,
+			txnMaps: nil,
+		}
+	}
+	return &txnScheduler{
+		latches: latch.NewLatches(txnSchedulerLatchesSize),
+		txnMaps: make([]txnMap, txnSchedulerMapsCount),
+	}
+}
+
+func (scheduler *txnScheduler) execute(ctx context.Context, txn *twoPhaseCommitter, connID uint64) error {
+	newCommitter := newTxnCommiter(ctx, txn)
+	// latches disabled
+	if scheduler.latches == nil {
+		_, err := newCommitter.execute()
+		return errors.Trace(err)
+	}
+	// latches enabled
+	newCommitter.lock = scheduler.latches.GenLock(txn.startTS, txn.keys)
+	// for transactions which is not retryable, commit directly.
+	if !sessionctx.GetRetryAble(ctx) {
+		err := scheduler.runForUnRetryAble(newCommitter)
+		return errors.Trace(err)
+	}
+	// for transactions which need to acquire latches
+	ch := make(chan error, 1)
+	newCommitter.ch = ch
```

Comment:
Why not create `ch` when create the `committer`?

Reply:
We do not `ch` when `latches` are disabled or transactions  are not retryable @disksing